src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonPropertyInfo.cs

```diff
@@ -654,11 +654,25 @@ public string Name
                 }
 
                 _name = value;
+
+                // This setter should only be called by end users.
+                // Disable fast-path if a user modifies a property name.
```

Comment:
If the user modifies source gen metadata, then they're necessarily using a contract resolver other than source generated `JsonSerializerContext` instances. In such cases I would expect that `CanUseSerializeHandler` is always flipped back to false:

https://github.com/dotnet/runtime/blob/bce23abb964d81b6d626f2a2b435549636769a1d/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Metadata/JsonTypeInfo.cs#L537

Is there any particular case that was missed out and there's a failing test without this check?

Reply:
That would be this:

```cs
JsonTypeInfo<Person> typeInfo = SourceGenContext.Default.Person;

foreach (JsonPropertyInfo property in typeInfo.Properties)
{
    property.Name = property.Name.ToUpperInvariant();
}

string json = JsonSerializer.Serialize(person, typeInfo);
JsonTestHelper.AssertJsonEqual(@"{""FIRSTNAME"":""Jane"",""LASTNAME"":""Doe""}", json); // Fail: regular-case is serialized instead (because fast-path is taken; metadata should be used instead since we have it).
Person person = JsonSerializer.Deserialize(json, typeInfo); 
Assert.Equal("Jane", person.FirstName);
Assert.Equal("Doe", person.LastName);
```

Reply:
Oh wow. Fundamentally I think the problem here is that static metadata is mutable even though these should be locked for modification. We should try to service this.

cc @krwq